pyanaconda/payload/requirement.py

```diff
@@ -0,0 +1,202 @@
+# the GNU General Public License v.2, or (at your option) any later version.
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY expressed or implied, including the implied warranties of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General
+# Public License for more details.  You should have received a copy of the
+# GNU General Public License along with this program; if not, write to the
+# Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+# 02110-1301, USA.  Any Red Hat trademarks that are incorporated in the
+# source code or documentation are not subject to the GNU General Public
+# License and may only be used or replicated with the express permission of
+# Red Hat, Inc.
+#
+
+from collections import OrderedDict, namedtuple
+from pyanaconda.core.constants import PayloadRequirementType
+from pyanaconda.payload.errors import PayloadRequirementsMissingApply
+
+from pyanaconda.anaconda_loggers import get_module_logger
+
+log = get_module_logger(__name__)
+
+PayloadRequirementReason = namedtuple('PayloadRequirementReason', ['reason', 'strong'])
+
+
+__all__ = ["PayloadRequirements", "PayloadRequirement"]
+
+
+class PayloadRequirement(object):
+    """An object to store a payload requirement with info about its reasons.
+
+    For each requirement multiple reasons together with their strength
+    can be stored in this object using the add_reason method.
+    A reason should be just a string with description (ie for tracking purposes).
+    Strength is a boolean flag that can be used to indicate whether missing the
+    requirement should be considered fatal. Strength of the requirement is
+    given by strength of all its reasons.
+    """
+    def __init__(self, req_id, reasons=None):
+        self._id = req_id
+        self._reasons = reasons or []
+
+    @property
+    def id(self):
+        """Identifier of the requirement (eg a package name)"""
+        return self._id
+
+    @property
+    def reasons(self):
+        """List of reasons for the requirement"""
+        return [reason for reason, strong in self._reasons]
+
+    @property
+    def strong(self):
+        """Strength of the requirement (ie should it be considered fatal?)"""
+        return any(strong for reason, strong in self._reasons)
+
+    def add_reason(self, reason, strong=False):
+        """Adds a reason to the requirement with optional strength of the reason"""
+        self._reasons.append(PayloadRequirementReason(reason, strong))
+
+    def __str__(self):
+        return "PayloadRequirement(id=%s, reasons=%s, strong=%s)" % (self.id,
+                                                                     self.reasons,
+                                                                     self.strong)
+
+    def __repr__(self):
+        return 'PayloadRequirement(id=%s, reasons=%s)' % (self.id, self._reasons)
+
+
+class PayloadRequirements(object):
```

Comment:
The support for declaring the requirements might be a good candidate for moving on DBus.

Reply:
True. It could be easy to move. Thanks for idea.